### PR TITLE
photoprism: 221102-905925b4d -> 221118-e58fee0fb

### PR DESCRIPTION
**Pull request by BeneSim**
###### Description of changes

221103-211eb36ea:  https://github.com/photoprism/photoprism/releases/tag/221103-211eb36ea
221104-20d180b21: https://github.com/photoprism/photoprism/releases/tag/221104-20d180b21
221105-7a295cab4:  https://github.com/photoprism/photoprism/releases/tag/221105-7a295cab4
221116-122ebfb70: https://github.com/photoprism/photoprism/releases/tag/221116-122ebfb70
221117-3268c4de8: https://github.com/photoprism/photoprism/releases/tag/221117-3268c4de8
221118-e58fee0fb: https://github.com/photoprism/photoprism/releases/tag/221118-e58fee0fb

Result of `nixpkgs-review`
```
2 packages marked as broken and skipped:
dat slack-dark

2 packages failed to build:
flexoptix-app imgbrd-grabber

84 packages built:
antora balanceofsatoshis bat-extras.prettybat bitwarden-cli castnow code-server commitlint concurrently create-cycle-app discourse discourseAllPlugins emanote emojione epgstation fast-cli gnomeExtensions.pop-shell google-clasp grafana-image-renderer gtop image_optim joplin lessc mastodon-bot node2nix openmoji-black openmoji-color photoprism pm2 postcss-cli pulp pyright quarto redoc-cli reveal-md slack sloc styx teams theLoungePlugins.plugins.closepms theLoungePlugins.plugins.giphy theLoungePlugins.themes.flat-blue theLoungePlugins.themes.flat-dark theLoungePlugins.themes.new-morning theLoungePlugins.themes.new-morning-compact thelounge typescript uppy-companion vimPlugins.YouCompleteMe vimPlugins.coc-clangd vimPlugins.coc-diagnostic vimPlugins.coc-git vimPlugins.coc-haxe vimPlugins.coc-html vimPlugins.coc-java vimPlugins.coc-json vimPlugins.coc-lists vimPlugins.coc-metals vimPlugins.coc-pairs vimPlugins.coc-prettier vimPlugins.coc-pyright vimPlugins.coc-rust-analyzer vimPlugins.coc-sh vimPlugins.coc-snippets vimPlugins.coc-spell-checker vimPlugins.coc-sqlfluff vimPlugins.coc-stylelint vimPlugins.coc-tsserver vimPlugins.coc-vetur vimPlugins.coc-vimlsp vimPlugins.markdown-preview-nvim vscode vscode-extensions.matklad.rust-analyzer vscode-extensions.ms-python.vscode-pylance vscode-fhs vscode-with-extensions vscodium vscodium-fhs wasm-text-gen weylus whalebird wrangler yaml-language-server ycmd zx
```
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by surfaceflinger** [comment]
@BeneSim I can't find any contact to you so I'm doing it through PRs :) Could you please share your config for NixOS container with photoprism? It's a pain to configure it for me due to nature of such apps. Thanks!

**Comment by Luflosi** [comment]
@surfaceflinger if you are looking for his contact info, his E-Mail address(es) is/are in the commit as the author and committer. GitHub doesn't seem to show it but if you open the git history with an external program, you can see it. It is also in the `maintainers/maintainer-list.nix` file.

If this program is so difficult to configure, I suggest adding some documentation when you figured it out.

**Comment by BeneSim** [comment]
> @BeneSim I can't find any contact to you so I'm doing it through PRs :) Could you please share your config for NixOS container with photoprism? It's a pain to configure it for me due to nature of such apps. Thanks!

Sorry for the late response. Guess this isn't really the right place to discuss this, but maybe other people are looking for this as well.

I'm afraid that my config is very minimal and I haven't developed a sophisticated module for photoprism yet. Using most of the default options with a mariadb database  in a nixos container would look something like this:

```nix
containers.photoprism = let
  mysqlDataDir = "/var/lib/mysql";
in
  {
  autoStart = true;

  bindMounts = {
    "/import" = {
      hostPath = "/host/path/to/import";
      isReadOnly = false;
    };
    "/originals" = {
      hostPath = "/host/path/to/photos";
      isReadOnly = false;
    };
    "${mysqlDataDir}" = {
      hostPath = "/host/path/to/mysql-data";
      isReadOnly = false;
    };
  };

  config = { config, pkgs, ... }: let
    photoprismUser = "photoprism";
    photoprismGroup = photoprismUser;
    photoprismDatabase = "photoprism";
  in
    {
    # If you want to use a more recent version of photoprism. Otherwise skip this overlay
    # and use the version in 22.11 or unstable. The rest of the config should be compatible.
    nixpkgs.overlays =
      let
        owner = "BeneSim";
        branchname = "photoprism-221102-905925b4d-221118-e58fee0fb";
        pkgsReview = pkgs.fetchzip {
          url = "https://github.com/${owner}/nixpkgs/archive/${branchname}.tar.gz";
          sha256 = "sha256-3Q5hGLn14DeszCtctjvsW6R/JAz1bD3RfxARKP3zTig=";
        };
      in
      [
        (self: super: {
          review = import pkgsReview { overlays = [ ]; config = super.config; };
          photoprism = self.review.photoprism;
        })
      ];

    # I'm using a rather complicated network setup (not shown here), if you don't use privateNetwork
    # you should add this to the "host config"
    # networking.firewall.allowedTCPPorts = [ 2342 ];

    users.groups."${photoprismGroup}" = { };

    users.users."${photoprismUser}" = {
      isSystemUser = true;
      group = photoprismGroup;
      home = "/var/lib/photoprism";
    };

    # Added photoprism to the environment to be able to use the photoprism cli
    environment.systemPackages = with pkgs; [
      photoprism
    ];

    systemd.services.photoprism = {
      wantedBy = [ "multi-user.target" ];
      description = "Photoprism";
      environment = {
        PHOTOPRISM_ADMIN_PASSWORD = "supersecret";
        PHOTOPRISM_DATABASE_DRIVER = "mysql";
        PHOTOPRISM_DATABASE_NAME = photoprismDatabase;
        PHOTOPRISM_DATABASE_SERVER = "/run/mysqld/mysqld.sock";
        PHOTOPRISM_DATABASE_USER = photoprismUser;
      };
      serviceConfig = {
        User = photoprismUser;
        Group = photoprismGroup;
        ExecStart = "${pkgs.photoprism}/bin/photoprism start";
      };
    };

    services.mysql = {
      enable = true;
      dataDir = mysqlDataDir;
      package = pkgs.mariadb;
      ensureUsers = [
        {
          name = photoprismUser;
          ensurePermissions = {
            "${photoprismDatabase}.*" = "ALL PRIVILEGES";
          };
        }
      ];
      ensureDatabases = [
        photoprismDatabase
      ];
    };
    system.stateVersion = "22.05";
  };
};

```

**Comment by BeneSim** [comment]
Regarding this PR:

The way the frontend is currently placed in nixpkgs is via the `node-packages.json` which had been suggested in (https://github.com/NixOS/nixpkgs/pull/170953#pullrequestreview-958624013). While this does work quite good and might be the best solution right now in theory, in practice this is really cumbersome. Running the `generate.sh` and `nixpkgs-review` takes more than a workday on my machine ... Not to mention the vast amount of lines of code changes the reviewers get confronted with in those kind of PRs.  Photoprism is evolving fast and I'd like to keep up with it.

I'd love to eject from `node-packages.json` and rather use something like `node2nix` instead (haven't dug into the details yet). Sure, this would mean that nixpkgs lines of code would grow a lot more (compared to the current approach). But it would be a whole lot more maintain- and reviewable.

What do you think about this?

**Comment by winterqt** [comment]
I completely forgot to PR this, but a better solution would be to use `buildNpmPackage`:

```diff
diff --git a/pkgs/servers/photoprism/frontend.nix b/pkgs/servers/photoprism/frontend.nix
index 0eb5731e222..9c03ccccb82 100644
--- a/pkgs/servers/photoprism/frontend.nix
+++ b/pkgs/servers/photoprism/frontend.nix
@@ -1,39 +1,20 @@
-{ lib, nodePackages, nodejs-14_x, stdenv, src, version, ... }:
+{ lib, buildNpmPackage, src, version }:
 
-let
-  nodeDependencies = nodePackages.photoprism-frontend.override {
-    inherit version;
-    name = "photoprism-frontend-dependencies";
-
-    # Workaround for lack of sourceRoot option in buildNodePackage.
-    src = "${src}/frontend";
-
-    meta.broken = false;
-  };
-in
-stdenv.mkDerivation {
+buildNpmPackage {
   inherit src version;
   pname = "photoprism-frontend";
 
-  buildInputs = [ nodejs-14_x ];
-
-  buildPhase = ''
-    runHook preBuild
-
-    pushd frontend
-    ln -s ${nodeDependencies}/lib/node_modules/photoprism/node_modules ./node_modules
-    export PATH="${nodeDependencies}/lib/node_modules/photoprism/node_modules/.bin:$PATH"
-    NODE_ENV=production npm run build
-    popd
-
-    runHook postBuild
+  postPatch = ''
+    cd frontend
   '';
 
+  npmDepsHash = lib.fakeHash;
+
   installPhase = ''
     runHook preInstall
 
     mkdir $out
-    cp -r assets $out/
+    cp -r ../assets $out/
 
     runHook postInstall
   '';
```

**Comment by BeneSim** [comment]
> I completely forgot to PR this, but a better solution would be to use `buildNpmPackage`, which puts all the dependencies in a FOD, making updates very quick in comparison to updating the nodePackages set, and reduces Nixpkgs code size:
> 
> ```diff
> diff --git a/pkgs/servers/photoprism/frontend.nix b/pkgs/servers/photoprism/frontend.nix
> index 0eb5731e222..9c03ccccb82 100644
> --- a/pkgs/servers/photoprism/frontend.nix
> +++ b/pkgs/servers/photoprism/frontend.nix
> @@ -1,39 +1,20 @@
> -{ lib, nodePackages, nodejs-14_x, stdenv, src, version, ... }:
> +{ lib, buildNpmPackage, src, version }:
>  
> -let
> -  nodeDependencies = nodePackages.photoprism-frontend.override {
> -    inherit version;
> -    name = "photoprism-frontend-dependencies";
> -
> -    # Workaround for lack of sourceRoot option in buildNodePackage.
> -    src = "${src}/frontend";
> -
> -    meta.broken = false;
> -  };
> -in
> -stdenv.mkDerivation {
> +buildNpmPackage {
>    inherit src version;
>    pname = "photoprism-frontend";
>  
> -  buildInputs = [ nodejs-14_x ];
> -
> -  buildPhase = ''
> -    runHook preBuild
> -
> -    pushd frontend
> -    ln -s ${nodeDependencies}/lib/node_modules/photoprism/node_modules ./node_modules
> -    export PATH="${nodeDependencies}/lib/node_modules/photoprism/node_modules/.bin:$PATH"
> -    NODE_ENV=production npm run build
> -    popd
> -
> -    runHook postBuild
> +  postPatch = ''
> +    cd frontend
>    '';
>  
> +  npmDepsHash = lib.fakeHash;
> +
>    installPhase = ''
>      runHook preInstall
>  
>      mkdir $out
> -    cp -r assets $out/
> +    cp -r ../assets $out/
>  
>      runHook postInstall
>    '';
> ```

Awesome! Applied the patch to my source tree, adjusted the hashDeps and it works flawlessly. The only question that remain from my point of view would be if we should commit this to another PR or use this PR instead? Guess it would be cleaner to use a different PR for that purpose. Would you mind creating that PR? I could issue a PR too if you would like me to.

**Comment by winterqt** [comment]
I think it's fine here -- there's already a conflict in `node-packages.nix`, so may be best to put it here anyways.

My main uncertainty is how to structure this. For example, should you just do `photoprism x -> y + convert to buildNpmPackage`, or split into two commits (with the `buildNpmPackage` one coming before the bump to not have conflicts + an unnecessary nodePackages generation)?

Thoughts?

**Comment by Stunkymonkey** [comment]
maybe someone can use my `photoprism` module: https://github.com/NixOS/nixpkgs/pull/207567

**Comment by BeneSim** [comment]
Sorry, I've been busy during the holidays. I switched to using `buildNpmPackage`  with the patch supplied by @winterqt (thanks again for that). Based on that I did the version upgrade.

There is one flaw with [3a849a8](https://github.com/NixOS/nixpkgs/pull/204521/commits/3a849a8c06d616d96e29c67b956603dacecb3d35) though, I wasn't able to run `pkgs/development/node-packages/generate.sh` with the new tree since it keeps failing with
```
fetching local directory: ../quicktype-core from quicktype/quicktype-graphql-input
Error: ENOENT: no such file or directory, open '/home/username/projects/nixpkgs/pkgs/development/node-packages/quicktype/quicktype-core/package.json
```
I'm going to create an issue for that.

The question remain wether we should wait for the issue to be resolved so that I can fixup [3a849a8](https://github.com/NixOS/nixpkgs/pull/204521/commits/3a849a8c06d616d96e29c67b956603dacecb3d35) by running `generate.sh`  and getting rid of everything related to photoprism or simply use it as is and wait for some other PR to run `generate.sh`?

I'm actually tired of leaving my pc on over night to run this `generate.sh`just to see that it failed half way :(

**Comment by BeneSim** [comment]
Issue regarding `quicktype` has been resolved upstream, see https://github.com/NixOS/nixpkgs/issues/211268 and https://github.com/quicktype/quicktype/issues/2144 . I regenerated the `node-packages.nix` and thus this PR should now be ready to get reviewed / merged.

File: pkgs/servers/photoprism/frontend.nix [hunk 1]
```diff
@@ -1,39 +1,20 @@
...
-    runHook preBuild
-
-    pushd frontend
-    ln -s ${nodeDependencies}/lib/node_modules/photoprism/node_modules ./node_modules
-    export PATH="${nodeDependencies}/lib/node_modules/photoprism/node_modules/.bin:$PATH"
-    NODE_ENV=production npm run build
-    popd
-
-    runHook postBuild
+  postPatch = ''
```
Comment (SuperSandro2000) [inline pkgs/servers/photoprism/frontend.nix:7]: This should really be preConfigure or preBuild

File: pkgs/servers/photoprism/frontend.nix [hunk 2]
```diff
@@ -8,7 +8,7 @@ buildNpmPackage {
     cd frontend
   '';
 
-  npmDepsHash = "sha256-NAtZ85WjiQn9w0B9Y78XL+tSsshHlaXS8+WtojFJmGg=";
+  npmDepsHash = "sha256-NAtZ85WjiQn9w0B9Y78XL+tSsshHlaXS8+WtojFJmGg";
```
Comment (winterqt) [inline pkgs/servers/photoprism/frontend.nix:11]: (Why was this done?)

**Review by winterqt — COMMENTED** [review]
Is there any reason the nodePackages generation was done in the bump commit, and not the commit that edited the node-packages.json file? Not that it matters now, since this was merged, but just curious.

**Comment by BeneSim** [comment]
> Is there any reason the nodePackages generation was done in the bump commit, and not the commit that edited the node-packages.json file? Not that it matters now, since this was merged, but just curious.

The reason is sloppiness on my side :( Same goes for the deletion of `=`in the `npmDepsHash` (will fix this on next version bump). Should've done a fix up of the first commit when `generate.sh` did finally work again for me. I rushed it and didn't give it a second thought, my bad.

**Pull request merged**